sound/soc/sof/pcm.c

```diff
@@ -254,13 +254,14 @@ static int sof_pcm_hw_free(struct snd_pcm_substream *substream)
 	if (!spcm)
 		return -EINVAL;
 
-	if (!spcm->prepared[substream->stream])
-		return 0;
-
 	dev_dbg(sdev->dev, "pcm: free stream %d dir %d\n", spcm->pcm.pcm_id,
 		substream->stream);
 
-	ret = sof_pcm_dsp_pcm_free(substream, sdev, spcm);
+	if (spcm->prepared[substream->stream]) {
+		ret = sof_pcm_dsp_pcm_free(substream, sdev, spcm);
+		if (ret < 0)
+			return ret;
```

Comment:
if ret < 0 you skip the follow-up cleanup...

Reply:
and the test below after cancel_work_sync may use an uninitialized value since ret is not set

Reply:
@plbossart Oops sorry, my bad, I was meaning to remove the check on L269-270. But, but, I just replied to https://github.com/thesofproject/linux/pull/1169 and it seems I have had incorrect assumptions about the FW ABI. Reflecting the same discussion here (and e.g. implementation of snd_pcm_hw_free() in pcm_native.c), the correct action would indeed seem to be capture the error but not give any of the following clean-up steps. I will update patch accordingly.